pkg/logs/tailers/windowsevent/tailer.go

```diff
@@ -89,210 +94,257 @@ func (t *Tailer) Identifier() string {
-	} else {
-		err = mv.SetValueForPath(dataField, dataPath)
-		if err != nil {
-			log.Debugf("Error formatting %s: %s", dataPath, err)
+// Start starts tailing the event log.
+func (t *Tailer) Start(bookmark string) {
+	log.Infof("Starting windows event log tailing for channel %s query %s", t.config.ChannelPath, t.config.Query)
+	t.stop = make(chan struct{})
+	t.done = make(chan struct{})
+	go t.forwardMessages()
+	t.decoder.Start()
+	go t.tail(bookmark)
+}
+
+// Stop stops the tailer
+func (t *Tailer) Stop() {
+	log.Info("Stop tailing windows event log")
+	close(t.stop)
+	<-t.done
+
+	t.decoder.Stop()
+
+	t.sub.Stop()
+}
+
+func (t *Tailer) forwardMessages() {
+	for decodedMessage := range t.decoder.OutputChan {
+		if len(decodedMessage.Content) > 0 {
+			t.outputChan <- decodedMessage
 		}
 	}
+}
 
-	// extract, parse then modify the Event.EventData.Binary data field
-	binaryData, err := extractParsedBinaryData(mv)
-	if err != nil {
-		log.Debugf("Error extracting binary data: %s", err)
+// tail subscribes to the channel for the windows events
+func (t *Tailer) tail(bookmark string) {
+	if t.evtapi == nil {
+		t.evtapi = winevtapi.New()
+	}
+
+	var err error
+
+	opts := []evtsubscribe.PullSubscriptionOption{
+		evtsubscribe.WithWindowsEventLogAPI(t.evtapi),
+		evtsubscribe.WithEventBatchCount(10),
+		evtsubscribe.WithNotifyEventsAvailable(),
+	}
+
+	if bookmark != "" {
+		// load bookmark
+		t.bookmark, err = evtbookmark.New(
+			evtbookmark.WithWindowsEventLogAPI(t.evtapi),
+			evtbookmark.FromXML(bookmark))
+		if err != nil {
+			err = fmt.Errorf("error creating bookmark: %v", err)
+			log.Errorf("%v", err)
+			t.source.Status.Error(err)
+			return
+		}
+		opts = append(opts, evtsubscribe.WithStartAfterBookmark(t.bookmark))
 	} else {
-		_, err = mv.UpdateValuesForPath("Binary:"+binaryData, binaryPath)
+		// new bookmark
+		t.bookmark, err = evtbookmark.New(
+			evtbookmark.WithWindowsEventLogAPI(t.evtapi))
 		if err != nil {
-			log.Debugf("Error formatting %s: %s", binaryPath, err)
+			err = fmt.Errorf("error creating bookmark: %v", err)
+			log.Errorf("%v", err)
+			t.source.Status.Error(err)
+			return
 		}
 	}
 
-	// Normalize the Event.System.EventID field
-	err = normalizeEventID(mv)
+	// subscription
+	t.sub = evtsubscribe.NewPullSubscription(
+		t.config.ChannelPath,
+		t.config.Query,
+		opts...,
+	)
+	err = t.sub.Start()
 	if err != nil {
-		log.Debugf("Error normalizing EventID: %s", err)
-	}
-
-	// Replace Task and Opcode codes by the rendered value
-	if re.task != "" {
-		_, _ = mv.UpdateValuesForPath("Task:"+re.task, taskPath)
-	}
-	if re.opcode != "" {
-		_, _ = mv.UpdateValuesForPath("Opcode:"+re.opcode, opcode)
-	}
-	// Set message and severity
-	if re.message != "" {
-		_ = mv.SetValueForPath(re.message, "message")
-	}
-	if re.level != "" {
-		_ = mv.SetValueForPath(re.level, "level")
+		err = fmt.Errorf("failed to start subscription: %v", err)
+		log.Errorf("%v", err)
+		t.source.Status.Error(err)
+		return
```

Comment:
The loop would attempt to restart the subscription if it fails after the start. Would it make sense to keep trying if it fails here as well?

Reply:
I was thinking it would be helpful to send the error to the user on startup in case they misconfigured something. But looking again I think if we log and set the `t.source.Status.Error` that will happen even if we continue and enter retry. I'll update it.

Reply:
https://github.com/DataDog/datadog-agent/pull/19761/commits/9d16fb9f87eef3c5e2789c2ccb668ac4fe31b08a